### PR TITLE
Stac 10665 topology events for snow

### PR DESCRIPTION
**Pull request by hruhek**
### Step 1: Link to Jira issue
https://stackstate.atlassian.net/browse/STAC-10665

### Step 2: Description of changes
Custom/Topology events for SNOW Change Requests.

### Step 3: Did you add / update tests for your changes in the right area?
- [x] Unit/Component test

### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Johnny

### Step 5: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

### CODE REVIEW

File: servicenow/stackstate_checks/servicenow/data/conf.yaml.example [hunk 1]
```diff
@@ -7,8 +7,14 @@ instances:
     user: admin
     password: Service@123
 
-    # Batch size for paginating results. Default value is 2500
-    # batch_size: 1000
+    # Batch size for paginating results. Default value is 2500. Max value is 10000.
+    # batch_size: 2500
+
+    # On first start we get all changed request that have been updated in last N days. Default value is 100.
+    # crs_updated_in_days: 30
```
Comment (vishalxebia) [inline servicenow/stackstate_checks/servicenow/data/conf.yaml.example:14]: we should change the name to something meaningful like `change_request_bootstrap_days` following the same structure as dynatrace. so everybody is on same page.

File: servicenow/stackstate_checks/servicenow/data/conf.yaml.example [hunk 2]
```diff
@@ -7,8 +7,14 @@ instances:
...
+
+    # Maximum number of change request that we process. Default value is 1000.
+    # crs_max_number: 200
```
Comment (vishalxebia) [inline servicenow/stackstate_checks/servicenow/data/conf.yaml.example:17]: should change to `change_request_process_limit`.

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 3]
```diff
@@ -167,73 +199,181 @@ def _process_components(self, components, instance_config):
...
+                old_state = crs_persisted_state.get(change_request.number)
+                if old_state is None or old_state != change_request.state:
+                    self._create_event_from_change_request(change_request, instance_info)
+                    crs_persisted_state[change_request.number] = change_request.state
+        self.persistent_state.set_state(self.persistent_instance, state)
+
+    def _load_state(self, instance_info):
+        """
+        Persistent State structure:
+        {
```
Comment (vishalxebia) [inline servicenow/stackstate_checks/servicenow/servicenow.py:311]: change the structure to 
{
      url: {
                  lastProcessedCRtimestamp: timestamp,
                  changeRequest: {
                                                      change_request_number: state
                                               }
             }

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 4]
```diff
@@ -167,73 +199,181 @@ def _process_components(self, components, instance_config):
...
+        :return: None
+        """
+        # TODO switch to schematics model
+        self.persistent_state = PersistentState()
+        start_dt = datetime.datetime.now() - datetime.timedelta(days=CRS_OLDEST_UPDATE_IN_NUMBER_OF_DAYS)
+        try:
+            self.persistent_state.get_state(self.persistent_instance)
+        except StateReadException:
+            self.log.info('First run! Creating new persistent state.')
+            empty_state = {
```
Comment (vishalxebia) [inline servicenow/stackstate_checks/servicenow/servicenow.py:329]: this empty state structure will change.

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 5]
```diff
@@ -167,73 +199,181 @@ def _process_components(self, components, instance_config):
...
+                instance_info.url: start_dt.strftime(TIME_FORMAT),
+                self.cr_persistence_key: {}
+            }
+            self.persistent_state.set_state(self.persistent_instance, data=empty_state)
+
+    def _create_event_from_change_request(self, change_request, instance_info):
+        cmdb_ci = change_request.cmdb_ci
+        identifiers = []
+        external_id = cmdb_ci['link'].split('/').pop()
```
Comment (vishalxebia) [inline servicenow/stackstate_checks/servicenow/servicenow.py:338]: If you are using `ALL` value for `sysparam_display` as discussed yesterday then you will not get `link` i think.

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 6]
```diff
@@ -167,73 +199,181 @@ def _process_components(self, components, instance_config):
...
+            'msg_title': msg_title,
+            'msg_text': msg_text,
+            # TODO do we need an aggregation_key?
+            # 'aggregation_key': '???',
+            'context': {
+                'source': 'servicenow',
+                'category': 'change_request',
+                'element_identifiers': identifiers,
+                'data': {
+                    'cmdb_ci': cmdb_ci,
```
Comment (vishalxebia) [inline servicenow/stackstate_checks/servicenow/servicenow.py:367]: it's a dictionary right ?

File: servicenow/stackstate_checks/servicenow/data/conf.yaml.example (hunk 1 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/data/conf.yaml.example:14]: Done.

File: servicenow/stackstate_checks/servicenow/data/conf.yaml.example (hunk 2 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/data/conf.yaml.example:17]: done

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 3 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:312]: done

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 4 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:330]: done

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 6 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:370]: yes, it is a dictionary. I really don't know if we need to pass `cmdb_ci`. 
We have `identifiers` with sys_id that is used as `external_id` and URN created from cmdb_ci `name`.

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 5 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:341]: That's true. Currently it gets me just `display_values` not `values`. I can set it to `ALL` and get them both but I will need to change `ChangeRequest(Model):` types and `_create_event_from_change_request` method that gets values from that model.

**Review by TizianoPerrucci — APPROVED** [review]

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 7]
```diff
@@ -117,123 +167,209 @@ def filter_empty_metadata(self, data):
...
+            if change_request.cmdb_ci:
+                if change_request.sys_updated_on > instance_info.state.latest_sys_updated_on:
+                    instance_info.state.latest_sys_updated_on = change_request.sys_updated_on
+                old_state = instance_info.state.change_requests.get(change_request.number)
+                if old_state is None or old_state != change_request.state:
+                    self._create_event_from_change_request(change_request)
+                    instance_info.state.change_requests[change_request.number] = change_request.state
+
+    def _create_event_from_change_request(self, change_request):
+        identifiers = [
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:335]: What are the values for `Service`, `Affected CIs` and `Impacted Services/CIs` in the jira ticket we expect them to be added here if present. Does it makes sense?

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 8]
```diff
@@ -117,123 +167,209 @@ def filter_empty_metadata(self, data):
...
+            'timestamp': timestamp,
+            'event_type': change_request.type,
+            'msg_title': msg_title,
+            'msg_text': change_request.description,
+            'context': {
+                'source': 'servicenow',
+                'category': 'change_request',
+                'element_identifiers': identifiers,
+                'source_links': [],
+                'data': {
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:363]: Missing `Service Offerings` here

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 9]
```diff
@@ -246,36 +382,27 @@ def _get_json_batch(self, url, offset, batch_size, timeout, auth):
...
+        if response.status_code != 200:
+            raise CheckException("Got %s when hitting %s" % (response.status_code, url))
+
         try:
-            resp = yaml.safe_load(resp.text.encode("utf-8"))
-        except Exception as e:
-            self.log.exception(str(e))
-            raise CheckException("Exception occurred while parsing response and the error is : {}".format(str(e)))
+            response_json = json.loads(response.text.encode('utf-8'))
+        except json_parse_exception as e:
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:393]: I think `response.text.encode('utf-8')` can throw an error that is not an json_parse_exception, we should also catch that

**Review by zandrewitte — CHANGES_REQUESTED** [review]

File: servicenow/requirements.in [hunk 10]
```diff
@@ -0,0 +1 @@
+schematics==2.1.0
```
[the remark this answers is not in the document]
Comment (hruhek) [reply servicenow/requirements.in:1]: removed

File: servicenow/stackstate_checks/servicenow/data/conf.yaml.example [hunk 11]
```diff
@@ -3,12 +3,18 @@ init_config:
...
+    user: username
+    password: secret
 
-    # Batch size for paginating results. Default value is 2500
-    # batch_size: 1000
+    # Batch size for paginating results. Default value is 2500. Max value is 10000.
+    # batch_size: 2500
+
+    # On first start we get all changed request that have been updated in last N days. Default value is 100.
+    # change_request_bootstrap_days: 30
```
[the remark this answers is not in the document]
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/data/conf.yaml.example:14]: changed

File: servicenow/stackstate_checks/servicenow/data/conf.yaml.example [hunk 12]
```diff
@@ -3,12 +3,18 @@ init_config:
...
+
+    # Maximum number of change request that we process. Default value is 1000.
+    # change_request_process_limit: 200
```
[the remark this answers is not in the document]
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/data/conf.yaml.example:17]: done

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 13]
```diff
@@ -117,123 +167,209 @@ def filter_empty_metadata(self, data):
...
             batch_number += 1
-            offset += instance_config.batch_size
+            offset += instance_info.batch_size
             self.log.info(
                 'Processed batch no. {} with {} items.'.format(batch_number, number_of_elements_in_current_batch)
             )
 
-    def _process_components(self, components, instance_config):
-        for component in components.get('result'):
+        return {'result': collection}
```
[the remark this answers is not in the document]
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:211]: I missed that part when I did batch refactoring. I'll fix that.

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 14]
```diff
@@ -117,123 +167,209 @@ def filter_empty_metadata(self, data):
...
+            'number:{}'.format(change_request.number),
+            'priority:{}'.format(change_request.priority),
+            'risk:{}'.format(change_request.risk),
+            'state:{}'.format(change_request.state),
+            'category:{}'.format(change_request.category),
+            'conflict_status:{}'.format(change_request.conflict_status),
+            'assigned_to:{}'.format(change_request.assigned_to)
+        ]
+
+        self.log.info('Creating event from CR: %s', change_request.number)
```
[the remark this answers is not in the document]
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py]: done

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 15]
```diff
@@ -117,123 +167,209 @@ def filter_empty_metadata(self, data):
...
+                if change_request.sys_updated_on > instance_info.state.latest_sys_updated_on:
+                    instance_info.state.latest_sys_updated_on = change_request.sys_updated_on
+                old_state = instance_info.state.change_requests.get(change_request.number)
+                if old_state is None or old_state != change_request.state:
+                    self._create_event_from_change_request(change_request)
+                    instance_info.state.change_requests[change_request.number] = change_request.state
+
+    def _create_event_from_change_request(self, change_request):
+        identifiers = [
+            change_request.cmdb_ci['value'],
```
[the remark this answers is not in the document]
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:342]: ` cmdb_ci = DictType(StringType, required=True)`
We need both of them.
`value` is `sys_id` and this is out `external_id`
`display_value` is CI name, and we use these for identifier.
Yes it should match.

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 16]
```diff
@@ -1,78 +1,124 @@
...
+
+class ChangeRequest(Model):
+    number = WrapperType(StringType, required=True, value_mapping=lambda x: x['display_value'])
+    state = WrapperType(StringType, required=True, value_mapping=lambda x: x['display_value'])
+    cmdb_ci = DictType(StringType, required=True)
+    sys_updated_on = WrapperType(DateTimeType, value_mapping=lambda x: x['value'])
+    business_service = WrapperType(StringType, value_mapping=lambda x: x['display_value'])
+    service_offering = WrapperType(StringType, value_mapping=lambda x: x['display_value'])
+    short_description = WrapperType(StringType, required=True, value_mapping=lambda x: x['display_value'])
+    description = WrapperType(StringType, default='No description', value_mapping=lambda x: x['display_value'])
```
Comment (hruhek) [inline servicenow/stackstate_checks/servicenow/servicenow.py:55]: default values are not working for `WrappedType`

Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:55]: I'm not sure that required works. @zandrewitte I'll need to your help with this.

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 8 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:356]: added service_offering

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 9 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:395]: Added new exception

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 16 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:54]: We fixed that :)

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 7 above)
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:328]: This is still on TODO list. :(
 I was pointed to table where I can get these values, but I have to test it to see if there is all data I need...

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 17]
```diff
@@ -117,123 +167,209 @@ def filter_empty_metadata(self, data):
...
+                identifiers.append(Identifiers.create_host_identifier(comp_name))
             identifiers.append(external_id)
             data.update(component)
-            data.update({"identifiers": identifiers, "tags": instance_config.instance_tags})
+            data.update({"identifiers": identifiers, "tags": instance_info.instance_tags})
 
             self.component(external_id, comp_type, data)
 
-    def _collect_relation_types(self, instance_config):
+    def _collect_relation_types(self, instance_info):
```
[the remark this answers is not in the document]
Comment (hruhek) [reply servicenow/stackstate_checks/servicenow/servicenow.py:235]: We got 40 relation types in response

File: servicenow/tests/test_servicenow.py [hunk 18]
```diff
@@ -343,23 +394,23 @@ def test_process_component_relations_with_sys_filter_change(self):
         Test _process_components to return whole topology when query changed in between
         """
         sys_class_filter = self.instance.get('include_resource_types')
-        instance_config.sys_class_filter = sys_class_filter
+        instance_info.sys_class_filter = sys_class_filter
         query_filter = self.check.get_sys_class_relation_filter_query(sys_class_filter)
-        expected_query = "sysparm_query=parent.sys_class_nameINcmdb_ci_netgear%2Ccmdb_ci_cluster%2Ccmdb_ci_app_server" \
-                         "%5Echild.sys_class_nameINcmdb_ci_netgear%2Ccmdb_ci_cluster%2Ccmdb_ci_app_server"
+        expected_query = 'parent.sys_class_nameINcmdb_ci_netgear,cmdb_ci_cluster,cmdb_ci_app_server' \
+                         '^child.sys_class_nameINcmdb_ci_netgear,cmdb_ci_cluster,cmdb_ci_app_server'
```
Comment (zandrewitte) [inline servicenow/tests/test_servicenow.py:400]: This is still encoded correctly when talking to ServiceNow API right?

**Review by zandrewitte — APPROVED** [review]

**Pull request merged**